network/omaha/server.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2015 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package omaha
+
+import (
+	"encoding/xml"
+	"net"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/coreos/mantle/Godeps/_workspace/src/gopkg.in/tylerb/graceful.v1"
+)
+
+func NewServer(addr string) (*Server, error) {
+	l, err := net.Listen("tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+
+	srv := &graceful.Server{
+		Server: &http.Server{
+			Addr: addr,
+		},
+	}
+
+	s := &Server{
+		Addr: addr,
+		l:    l,
+		srv:  srv,
+	}
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("/v1/update/", s.UpdateHandler)
+
+	s.srv.Handler = mux
+
+	return s, nil
+}
+
+type Server struct {
+	Addr string
+	l    net.Listener
+	srv  *graceful.Server
+
+	reqChanLock sync.Mutex
+	reqChan     chan *Request
+}
+
+func (s *Server) Serve() error {
+	err := s.srv.Serve(s.l)
+	if err != nil {
+		netErr, ok := err.(*net.OpError)
+		if ok {
+			if netErr.Err.Error() == "use of closed network connection" {
```

Comment:
yes, unfortunately this error is not typed nor is the instance of the error exported for comparison.

```
net/net.go:284:     errClosing                = errors.New("use of closed network connection")
```